ext_localconf.php

```diff
@@ -70,22 +32,63 @@
+    $GLOBALS['TYPO3_CONF_VARS']['SC_OPTIONS']['cms/layout/class.tx_cms_layout.php']['tt_content_drawItem'][] =
+        \B13\Container\Hooks\DrawItem::class;
+
+    // register icons
+    $GLOBALS['TYPO3_CONF_VARS']['SC_OPTIONS']['GLOBAL']['extTablesInclusion-PostProcessing']['tx_container'] =
+        \B13\Container\Hooks\TableConfigurationPostProcessing::class;
+
+    // remove container colPos from "unused" page-elements
+    $GLOBALS['TYPO3_CONF_VARS']['SC_OPTIONS']['cms/layout/class.tx_cms_layout.php']['record_is_used']['tx_container'] =
+        \B13\Container\Hooks\UsedRecords::class . '->addContainerChildren';
+
+    // add tx_container_parent parameter to wizard items
+    $GLOBALS['TYPO3_CONF_VARS']['SC_OPTIONS']['cms']['db_new_content_el']['wizardItemsHook']['tx_container'] =
+        \B13\Container\Hooks\WizardItems::class;
+
+    $commandMapHooks = [
+        'tx_container-post-process' => \B13\Container\Hooks\Datahandler\CommandMapPostProcessingHook::class,
+        'tx_container-before-start' => \B13\Container\Hooks\Datahandler\CommandMapBeforeStartHook::class,
+        'tx_container-delete' => \B13\Container\Hooks\Datahandler\DeleteHook::class,
+        'tx_container-after-finish' => \B13\Container\Hooks\Datahandler\CommandMapAfterFinishHook::class
+    ];
+
+    $datamapHooks = [
+        'tx_container-before-start' => \B13\Container\Hooks\Datahandler\DatamapBeforeStartHook::class,
+        'tx_container-after-database-operations' => \B13\Container\Hooks\Datahandler\DatamapAfterDatabaseOperationHook::class
+    ];
+
     // EXT:content_defender
     $packageManager = \TYPO3\CMS\Core\Utility\GeneralUtility::makeInstance(\TYPO3\CMS\Core\Package\PackageManager::class);
     if ($packageManager->isPackageActive('content_defender')) {
-        $GLOBALS['TYPO3_CONF_VARS']['SC_OPTIONS']['cms']['db_new_content_el']['wizardItemsHook']['tx_container-content-defender'] =
-            \B13\Container\ContentDefender\Hooks\WizardItems::class;
-        $GLOBALS['TYPO3_CONF_VARS']['SYS']['formEngine']['formDataGroup']['tcaDatabaseRecord'][ \B13\Container\ContentDefender\Form\FormDataProvider\TcaCTypeItems::class] = [
-            'depends' => [
-                \TYPO3\CMS\Backend\Form\FormDataProvider\TcaSelectItems::class
-            ]
-        ];
-
-        // must be after tx_container
-        $GLOBALS['TYPO3_CONF_VARS']['SC_OPTIONS']['t3lib/class.t3lib_tcemain.php']['processCmdmapClass']['tx_container-content-defender'] =
-            \B13\Container\ContentDefender\Hooks\CommandMapHook::class;
-
-        $GLOBALS['TYPO3_CONF_VARS']['SC_OPTIONS']['t3lib/class.t3lib_tcemain.php']['processDatamapClass']['tx_container-content-defender'] =
-            \B13\Container\ContentDefender\Hooks\DatamapHook::class;
+        if (version_compare($packageManager->getPackage('content_defender')->getPackageMetaData()->getVersion(), '3.1.0', '<')) {
+            trigger_error('update EXT:content_defender to 3.1', E_USER_DEPRECATED);
```

Comment:
this would be executed on EVERY TYPO3 Request. Do we really want that? I'd say we leave a silent bw compat and we're done with it.

Reply:
mh, why? Update EXT:content_defender immediately and you are done. The core itselfs triggers also some deprecations on every Request.